packages/component/src/Transcript/ActivityRow.tsx

```diff
@@ -65,6 +65,47 @@ const ActivityRow = forwardRef<HTMLLIElement, ActivityRowProps>(({ activity, chi
     [bodyRef, children]
   );
 
+  const activityIdRef = useRefFrom(activity.id);
+
+  const handleFormData = useCallback(
+    (event: FormDataEvent & { target: HTMLFormElement }) => {
+      const { webchatIncludeActivityId, webchatIncludeActivityKey } = event.target.dataset;
+      if (webchatIncludeActivityId) {
+        event.formData.set(webchatIncludeActivityId, activityIdRef.current ?? '');
+      }
+      if (webchatIncludeActivityKey) {
+        event.formData.set(webchatIncludeActivityKey, activityKeyRef.current);
+      }
+    },
+
+    [activityKeyRef, activityIdRef]
+  );
+
+  const prevArticleRef = useRef<HTMLElement>(null);
+
+  const wrappedRef = useCallback(
+    (el: HTMLElement | null) => {
+      if (prevArticleRef.current) {
+        prevArticleRef.current.removeEventListener('formdata', handleFormData);
+      }
+
+      if (el) {
+        el.addEventListener('formdata', handleFormData);
+      }
+
+      prevArticleRef.current = el;
+
+      if (ref) {
+        if (typeof ref === 'function') {
+          ref(el);
+        } else {
+          ref.current = el;
+        }
+      }
+    },
+    [handleFormData, ref]
+  );
+
```

Comment:
Verified and no useEffect needed since:
1. handleFormData is stable
2. Element changes are handled by ref callback
3. Cleanup on unmount is handled by ref callback